remappers/hostmetrics/filesystem.go

```diff
@@ -0,0 +1,151 @@
+				if _, exists := deviceMetrics[key]; !exists {
+					deviceMetrics[key] = &DeviceMetrics{}
+				}
+				dmetrics := deviceMetrics[key]
+				if state, ok := dp.Attributes().Get("state"); ok {
+					switch state.Str() {
+					case "used":
+						if metric.Name() == "system.filesystem.usage" {
+							dmetrics.TotalUsage += value
+							dmetrics.UsedBytes += value
+							addFileSystemMetrics(out, timestamp, dataset, "system.filesystem.used.bytes", device.Str(), mpoint.Str(), fstype.Str(), value)
+						} else {
+							dmetrics.TotalInodeUsage += value
+						}
+					case "free":
+						if metric.Name() == "system.filesystem.usage" {
+							dmetrics.TotalUsage += value
+							addFileSystemMetrics(out, timestamp, dataset, "system.filesystem.free", device.Str(), mpoint.Str(), fstype.Str(), value)
+							addFileSystemMetrics(out, timestamp, dataset, "system.filesystem.available", device.Str(), mpoint.Str(), fstype.Str(), value)
+						} else {
+							dmetrics.TotalInodeUsage += value
+							addFileSystemMetrics(out, timestamp, dataset, "system.filesystem.free_files", device.Str(), mpoint.Str(), fstype.Str(), value)
+						}
+					}
+				}
+			}
+
+		}
+	}
+
+	for key, dmetrics := range deviceMetrics {
+		device, mpoint, fstype := key.device, key.mpoint, key.fstype
+		if dmetrics.TotalUsage > 0 {
+			addFileSystemMetrics(out, timestamp, dataset, "system.filesystem.total", device, mpoint, fstype, dmetrics.TotalUsage)
+			usedPercentage := float64(dmetrics.UsedBytes) / float64(dmetrics.TotalUsage)
+			addFileSystemMetrics(out, timestamp, dataset, "system.filesystem.used.pct", device, mpoint, fstype, usedPercentage)
+		}
+		if dmetrics.TotalInodeUsage > 0 {
+			addFileSystemMetrics(out, timestamp, dataset, "system.filesystem.files", device, mpoint, fstype, dmetrics.TotalInodeUsage)
+		}
+
+	}
+	return nil
+}
+
+func addFileSystemMetrics[T number](out pmetric.MetricSlice,
+	timestamp pcommon.Timestamp,
+	dataset, name, device, mpoint, fstype string,
+	value T,
+) {
+	var intValue *int64
+	var doubleValue *float64
+	if i, ok := any(value).(int64); ok {
+		intValue = &i
+	} else if d, ok := any(value).(float64); ok {
+		doubleValue = &d
+	}
+
+	remappers.AddMetrics(out, dataset,
+		func(dp pmetric.NumberDataPoint) {
+			dp.Attributes().PutStr("system.filesystem.device_name", device)
+			dp.Attributes().PutStr("system.filesystem.mount_point", mpoint)
+			dp.Attributes().PutStr("system.filesystem.type", fstype)
+		},
+		remappers.Metric{
+			DataType:    pmetric.MetricTypeSum,
+			Name:        name,
+			Timestamp:   timestamp,
+			IntValue:    intValue,
+			DoubleValue: doubleValue,
+		},
+	)
+
+}
+
+func getAttributes(dp pmetric.NumberDataPoint) (device, mpoint, fstype pcommon.Value, ok bool) {
+	device, ok = dp.Attributes().Get("device")
+	if !ok {
+		return
+	}
+	mpoint, ok = dp.Attributes().Get("mountpoint")
+	if !ok {
+		return
+	}
+	fstype, ok = dp.Attributes().Get("type")
+
+	return
+}
```

Comment:
```go
func getDeviceKey(dp pmetric.NumberDataPoint) (d deviceKey, ok bool) {
```

Can we directly return device key from this function and rename this? There is no point in keeping this generic function here.